File: 846-hand-of-straights/main.go
package main

func isNStraightHand(hand []int, groupSize int) bool {

    if len(hand) % groupSize != 0 {
        return false
    }

    sort.Slice(hand, func(i, j int) bool {
        return hand[i] < hand[j]
    })

    count := make(map[int]int)
    for _, card := range hand {
        count[card]++
    }

    for _, card := range hand {
        if count[card] > 0 {
            for i := range(groupSize) {
                if count[card + i] > 0 {
                    count[card + i] -= 1
                }else {
                    return false
                }
            }
        }
    }

    return true

}

File: 846-hand-of-straights/main.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_n_straight_hand(hand: Vec<i32>, group_size: i32) -> bool {

        if &hand.len() % group_size as usize  != 0 {
            return false;
        }

        let mut count = HashMap::new();
        for &card in &hand {
            *count.entry(card).or_insert(0) += 1;
        }

        let mut sorted_hand = hand.clone();
        sorted_hand.sort();

        for &card in &sorted_hand {
            if let Some(&card_count) = count.get(&card) {
                if card_count > 0 {
                    for i in 0..group_size {
                        let current_card = card + i;
                        if let Some(count) = count.get_mut(&current_card) {
                            if *count > 0 {
                                *count -= 1;
                            } else {
                                return false;
                            }
                        } else {
                            return false;
                        }
                    }
                }
            }
        }

        true
    }
}
